Recycling command buffers (dynamic/recycling_command_buffers.md) (#86)
chuigda/Vulkan-Tutorial-Rust-CN: book/src/dynamic/recycling_command_buffers.md

> # 重用指令缓冲

Reviewer: 重用/复用/（指令缓冲的）再利用？以下 ditto

Author: 先用“重用”吧。说实话这章 Kyle Mayes 的思路不是特别好，太拘泥于内存的 recycling 了，并且很多东西甚至不能说是 reuse，而是 free 了再 create 个新的。不过我们下一轮再解决这个问题。

Note that we do need to be careful about when we call `update_command_buffer`. This method will reset the command buffer which could cause serious issues if the command buffer is still being used to render a previously submitted frame. This issue was also discussed in the [`Descriptor set layout and buffer` chapter](../uniform/descriptor_set_layout_and_buffer.html#updating-uniform-data) which is why the call to `App::update_uniform_buffer` is where it is. As discussed in more detail in that chapter, both of these calls only happen after the call to `wait_for_fences` which waits for the GPU to be done with the acquired swapchain image and its associated resources so we are safe to do whatever we want with the command buffer.

> 我们要注意，在调用 `update_command_buffer` 时要多加小心。如果一个指令缓冲仍然在渲染一个之前已经提交的帧，重置这个指令缓冲会引发严重的问题。这个问题在 [`Descriptor set layout and buffer` 章节](../uniform/descriptor_set_layout_and_buffer.html#updating-uniform-data)中也有讨论，这也是为什么我们在这里调用 `App::update_uniform_buffer`。正如在那一章中详细讨论的那样，这两个函数都是在调用 `wait_for_fences` 之后被调用的，`wait_for_fences` 会等待 GPU 用完获取到的交换链图像及其相关资源，所以我们可以放心地对指令缓冲做任何事情。

Author: 问题不大，第一版先这样，后面再调整

You might recall we've used this flag before, there should already be a usage of this flag in the `begin_single_time_commands` function. This flag isn't required by Vulkan for correctness if you are only using command buffers once before resetting or freeing them, but this knowledge may allow the Vulkan driver to better optimize its handling of our single-use command buffers.

你可能还记得我们之前在 `begin_single_time_commands` 函数中使用过这个标志。Vulkan 并不强制要求使用这个标志，但如果你只在重置或释放指令缓冲之前使用一次指令缓冲，有了这个标志提供的信息，Vulkan 驱动或许能更好地优化对一次性指令缓冲的处理。

### 2. 重新分配指令缓冲

Next we'll take a look at allocating new command buffers each frame.

接着我们来看看每一帧都重新分配指令缓冲要怎么做。

Replace the code used to reset the command buffer at the beginning of `update_command_buffer` with code that replaces the previous command buffer with a new command buffer.

> 将 `update_command_buffer` 开头用来重置指令缓冲的代码替换为用新的指令缓冲替换之前的指令缓冲的代码。

Reviewer: 这句话绕晕了

Author: 现在再读下试试，没问题了就点 Resolve（

You could now run the program and see that the program works exactly like it did before, but if you do don't leave it running for too long! You may have already noticed that we aren't freeing the previous command buffer before we allocate a new one. If you observe the memory usage of our program after this change you'll see the memory usage start rising alarmingly fast as we rapidly collect thousands of derelict command buffers that are never recycled.

现在你可以运行程序，你会发现程序的运行效果和之前完全一样，但是如果你不停地运行它，就会发现问题了！你可能已经注意到了，在分配新的指令缓冲之前，我们并没有释放之前的指令缓冲。如果你观察这个修改后的程序的内存使用情况，你会发现内存使用量会迅速上升，因为很快就会出现数千个被遗弃的指令缓冲，而这些指令缓冲永远不会被重用。

Return the memory used by the previous command buffer to the command pool by freeing it at the beginning of `update_command_buffer`.

> 在 `update_command_buffer` 的开头，将原先指令缓冲使用的内存返还给指令池，然后释放它。

Reviewer 2: ```suggestion
在 `update_command_buffer` 的开头释放原先的指令缓冲，将其使用的内存返还给指令池。
```

Now when you run the program you should see stable memory usage instead of the program trying to gobble up all of the RAM on your system as if it thinks it's an Electron application.

现在运行程序，你会发现内存使用量稳定了下来，而不是像 Electron 应用一样试图吞噬系统上所有的内存。

We no longer need the `vk::CommandPoolCreateFlags::RESET_COMMAND_BUFFER` flag for our command pool since we aren't resetting command pools any more. Leaving this flag wouldn't affect the correctness of our program, but it could have a negative performance impact since it forces the command pool to allocate command buffers in such a way that they are resettable.

现在我们不再需要 `vk::CommandPoolCreateFlags::RESET_COMMAND_BUFFER` 标志了，因为我们不再重置指令池了。保留这个标志不会影响我们程序的正确性，但是它可能会对性能产生负面影响，因为它会强制指令池以可以重置的方式分配指令缓冲。

We'll replace this flag with `vk::CommandPoolCreateFlags::TRANSIENT` which tells Vulkan that the command buffers we'll be allocating with this command pool will be "transient", i.e. short-lived.

> 我们把标志换成 `vk::CommandPoolCreateFlags::TRANSIENT`，这会告诉 Vulkan，我们将使用这个指令池分配的指令缓冲是“短暂的”，也就是说，它会非常短命。

Reviewer 2: ```suggestion
我们把标志换成 `vk::CommandPoolCreateFlags::TRANSIENT`，这会告诉 Vulkan，我们将使用这个指令池分配的指令缓冲是“稍纵即逝的”，也就是说，这些指令缓冲会非常短命。
```

Like `vk::CommandBufferUsageFlags::ONE_TIME_SUBMIT`, this flag does not affect the correctness of our program but it may allow the Vulkan driver to better optimize the handling of our short-lived command buffers.

> 和 `vk::CommandBufferUsageFlags::ONE_TIME_SUBMIT` 一样，这个标志不会影响我们程序的正确性，但是它可能会让 Vulkan 驱动更好地优化对我们短命的指令缓冲的处理。

Reviewer 2: ```suggestion
和 `vk::CommandBufferUsageFlags::ONE_TIME_SUBMIT` 一样，这个标志不会影响我们程序的正确性，但是它可能会让 Vulkan 驱动更好地优化对短命的指令缓冲的处理。
```

### 3. 重置指令池

Next we'll look at resetting our entire command pool which will reset all of our active command buffers in one fell swoop.

> 接着我们来看看重置整个指令池，这会一举重置我们所有的活动指令缓冲。

Reviewer 2: ```suggestion
接着我们来看看重置整个指令池，这会一举重置所有存活的指令缓冲。
```

However, we immediately run into a problem with this approach. We can't reset *all* of our command buffers each frame because some of them might still be in use! The `wait_for_fences` call in `App::render` ensures that we are safe to reset the command buffer associated with the current framebuffer, but there might be other command buffers still in use.

然而，我们立刻就遇到了一个问题。我们不能每一帧都重置*所有*的指令缓冲，因为有些指令缓冲可能仍然在使用中！`App::render` 中的 `wait_for_fences` 调用确保我们可以安全地重置当前帧缓冲的指令缓冲，但是可能还有其他指令缓冲仍然在使用中。

We could continue down this path, but it would prevent our program from having multiple frames in-flight concurrently. This ability is important to maintain because, as discussed back in the [`Rendering and presentation` chapter](../drawing/rendering_and_presentation.html#frames-in-flight), it allows us to better leverage our hardware since the CPU will spend less time waiting on the GPU and vice-versa.

> 我们可以继续沿着这条路走下去，但这样的话我们的程序就不能并行渲染多个帧了。保持这种能力很重要，因为正如在 [`Rendering and presentation` 章节](../drawing/rendering_and_presentation.html#frames-in-flight)中提到的，这可以让我们更好地利用硬件，因为 CPU 会花费更少的时间等待 GPU，反之亦然。

Reviewer 2: ```suggestion
我们可以继续沿着这条路走下去，但这样的话我们的程序就不能并行渲染多个帧了。保持多帧并行渲染的能力很重要，因为正如在 [渲染与呈现](../drawing/rendering_and_presentation.html#frames-in-flight)那一章里提到的，这可以让我们更好地利用硬件，因为 CPU 会花费更少的时间等待 GPU，反之亦然。
```

Finally, delete the call to `free_command_buffers` in `App::destroy_swapchain`. This call now incorrectly attempts to return the memory assigned to the per-framebuffer command buffers to the global command pool despite the fact that these command buffers are no longer allocated from this command pool. Leaving this code in will most likely result in our program crashing when resizing the window or otherwise forcing a recreation of the swapchain. We no longer need to manage the deletion of individual command buffers since we are now managing this at the command pool level.

> 最后，在 `App::destroy_swapchain` 中删除对 `free_command_buffers` 的调用。这个调用现在会错误地尝试将分配给每帧指令缓冲的内存返还给全局指令池，尽管这些指令缓冲不再是从这个指令池分配的。如果保留这段代码，当窗口大小改变或者强制重新创建交换链时，我们的程序很可能会崩溃。我们不再需要管理单个指令缓冲的删除，因为我们现在是在指令池级别管理这个。

Reviewer 2: ```suggestion
最后，在 `App::destroy_swapchain` 中删除对 `free_command_buffers` 的调用。这个调用会错误地尝试将分配给每帧指令缓冲的内存返还给全局指令池，而现在这些指令缓冲并不是从全局指令池指令池分配的。如果保留这段代码，当窗口大小改变或者强制重新创建交换链时，我们的程序很可能会崩溃。我们不再需要管理单个指令缓冲的删除，因为我们现在是在指令池级别管理这个。
```
